tests/test_xxd/xxd.patched.1
.TH XXD 1 "August 1996" "Manual page for xxd"
.\"
.\" 25th May 1996
.\" Man page author:
.\"    Tony Nugent <tony@sctnugen.ppp.gu.edu.au> <T.Nugent@sct.gu.edu.au>
.\"    Changes by Bram Moolenaar <Bram@vim.org>
.SH NAME
.I xxd
\- make a hexdump or do the reverse.
.SH SYNOPSIS
.B xxd
\-h[elp]
.br
.B xxd
[options] [infile [outfile]]
.br
.B xxd
\-r[evert] [options] [infile [outfile]]
.SH DESCRIPTION
.I xxd
creates a hex dump of a given file or standard input.
It can also convert a hex dump back to its original binary form.
Like
.BR uuencode(1)
and
.BR uudecode(1)
it allows the transmission of binary data in a `mail-safe' ASCII representation,
but has the advantage of decoding to standard output.
Moreover, it can be used to perform binary file patching.
.SH OPTIONS
If no
.I infile
is given, standard input is read.
If
.I infile
is specified as a
.RB \` \- '
character, then input is taken from standard input.
If no
.I outfile
is given (or a
.RB \` \- '
character is in its place), results are sent to standard output.
.PP
Note that a "lazy" parser is used which does not check for more than the first
option letter, unless the option is followed by a parameter.
Spaces between a single option letter and its parameter are optional.
Parameters to options can be specified in decimal, hexadecimal or octal
notation.
Thus
.BR \-c8 ,
.BR "\-c 8" ,
.B \-c 010
and
.B \-cols 8
are all equivalent.
.PP
.TP
.IR \-a " | " \-autoskip
toggle autoskip: A single '*' replaces nul-lines.  Default off.
.TP
.IR \-b " | " \-bits
Switch to bits (binary digits) dump, rather than hexdump.
This option writes octets as eight digits "1"s and "0"s instead of a normal
hexacecimal dump. Each line is preceded by a line number in hexadecimal and
followed by an ascii (or ebcdic) representation. The command line switches
\-r, \-p, \-i do not work with this mode.
.TP
.IR "\-c cols " | " \-cols cols"
.IR "\-c cols " | " \-cols cols"
format
.RI < cols >
octets per line. Default 16 (\-i: 12, \-ps: 30, \-b: 6). Max 256.
.TP
.IR \-E " | " \-EBCDIC
Change the character encoding in the righthand column from ASCII to EBCDIC.
This does not change the hexadecimal representation. The option is
meaningless in combinations with \-r, \-p or \-i.
.TP
.IR "\-g bytes " | " \-groupsize bytes"
seperate the output of every
.RI < bytes >
bytes (two hex characters or eight bit-digits each) by a whitespace.
Specify
.I \-g 0
to suppress grouping.
.RI < Bytes "> defaults to " 2
in normal mode and \fI1\fP in bits mode.
Grouping does not apply to postscript or include style.
.TP
.IR \-h " | " \-help
print a summary of available commands and exit.  No hex dumping is performed.
.TP
.IR \-i " | " \-include
output in C include file style. A complete static array definition is written
(named after the input file), unless xxd reads from stdin.
.TP
.IR "\-l len " | " \-len len"
stop after writing
.RI  < len >
octets.
.TP
.IR \-p " | " \-ps " | " \-postscript " | " \-plain
output in postscript continuous hexdump style. Also known as plain hexdump
style.
.TP
.IR \-r " | " \-revert
reverse operation: convert (or patch) hexdump into binary.
If not writing to stdout, xxd writes into its output file without truncating
it. Use the combination
.I \-r \-p
to read plain hexadecimal dumps without line number information and without a
particular column layout. Additional Whitespace and line-breaks are allowed
anywhere.
.TP
.I \-seek offset
When used after
.I \-r
: revert with
.RI < offset >
added to file positions found in hexdump.
.TP
.I \-s [\+][\-]seek
start at
.RI < seek >
bytes abs. (or rel.) infile offset.
\fI\+ \fRindicates that the seek is relative to the current stdin file position
(meaningless when not reading from stdin).  \fI\- \fRindicates that the seek
should be that many characters from the end of the input (or if combined with
\fI \+ \fR: before the current stdin file position).
Without \-s option, xxd starts at the current file position.
.TP
.I \-u
use upper case hex letters. Default is lower case.
.TP
.IR \-v " | " \-version
show version string.
.SH CAVEATS
.PP
.I xxd \-r
has some builtin magic while evaluating line number information.
If the ouput file is seekable, then the linenumbers at the start of each
hexdump line may be out of order, lines may be missing, or overlapping. In
these cases xxd will lseek(2) to the next position. If the output file is not
seekable, only gaps are allowed, which will be filled by null-bytes.
.PP
.I xxd \-r
never generates parse errors. Garbage is silently skipped.
.PP
When editing hexdumps, please note that
.I xxd \-r
skips everything on the input line after reading enough columns of hexadecimal
data (see option \-c). This also means, that changes to the printable ascii (or
ebcdic) columns are always ignored. Reverting a plain (or postscript) style
hexdump with xxd \-r \-p does not depend on the correct number of columns. Here an thing that looks like a pair of hex-digits is interpreted.
.PP
Note the difference between
.br
\fI% xxd \-i file\fR
.br
and
.br
\fI% xxd \-i \< file\fR
.PP
.I xxd \-s \+seek
may be different from
.I xxd \-s seek
, as lseek(2) is used to "rewind" input.  A '+'
makes a difference if the input source is stdin, and if stdin's file position
is not at the start of the file by the time xxd is started and given its input.
The following examples may help to clarify (or further confuse!)...
.PP
Rewind stdin before reading; needed because the `cat' has already read to the
end of stdin.
.br
\fI% sh \-c 'cat > plain_copy; xxd \-s 0 > hex_copy' < file
.PP
Hexdump from file position 0x480 (=1024+128) onwards.
The `+' sign means "relative to the current position", thus the `128' adds to
the 1k where dd left off.
.br
\fI% sh \-c 'dd of=plain_snippet bs=1k count=1; xxd \-s +128 > hex_snippet' < file
.PP
Hexdump from file position 0x100 ( = 1024-768) on.
.br
\fI% sh \-c 'dd of=plain_snippet bs=1k count=1; xxd \-s +-768 > hex_snippet' < file
.PP
However, this is a rare situation and the use of `+' is rarely needed.
the author prefers to monitor the effect of xxd with strace(1) or truss(1), whenever \-s is used.
.SH EXAMPLES
.PP
.br
Print everything but the first three lines (hex 0x30 bytes) of
.B file
\.
.br
\fI% xxd \-s 0x30 file
.PP
.br
Print 3 lines (hex 0x30 bytes) from the end of
.B file
\.
.br
\fI% xxd \-s \-0x30 file
.PP
.br
Print 120 bytes as continuous hexdump with 40 octets per line.
.br
\fI% xxd \-l 120 \-ps \-c 20 xxd.1\fR
.br
2e544820585844203120224d616e75616c207061
.br
676520666f7220787864220a2e5c220a2e5c2220
.br
32317374204d617920313939360a2e5c22204d61
.br
6e207061676520617574686f723a0a2e5c222020
.br
2020546f6e79204e7567656e74203c746f6e7940
.br
7363746e7567656e2e7070702e67752e6564752e
.br

.br
Hexdump the first 120 bytes of this man page with 12 octets per line.
.br
\fI% xxd \-l 120 \-c 12 xxd.1\fR
.br
0000000: 2e54 4820 5858 4420 3120 224d  .TH XXD 1 "M
.br
000000c: 616e 7561 6c20 7061 6765 2066  anual page f
.br
0000018: 6f72 2078 7864 220a 2e5c 220a  or xxd"..\\".
.br
0000024: 2e5c 2220 3231 7374 204d 6179  .\\" 21st May
.br
0000030: 2031 3939 360a 2e5c 2220 4d61   1996..\\" Ma
.br
000003c: 6e20 7061 6765 2061 7574 686f  n page autho
.br
0000048: 723a 0a2e 5c22 2020 2020 546f  r:..\\"    To
.br
0000054: 6e79 204e 7567 656e 7420 3c74  ny Nugent <t
.br
0000060: 6f6e 7940 7363 746e 7567 656e  ony@sctnugen
.br
000006c: 2e70 7070 2e67 752e 6564 752e  .ppp.gu.edu.
.PP
.br
Display just the date from the file xxd.1
.br
\fI% xxd \-s 0x28 \-l 12 \-c 12 xxd.1\fR
.br
0000028: 3231 7374 204d 6179 2031 3939  21st May 199
.PP
.br
Copy
.B input_file
to
.B output_file
and prepend 100 bytes of value 0x00.
.br
\fI% xxd input_file | xxd \-r \-s 100 \> output_file\fR
.br

.br
Patch the date in the file xxd.1
.br
\fI% echo '0000029: 3574 68' | xxd \-r \- xxd.1\fR
.br
\fI% xxd \-s 0x28 \-l 12 \-c 12 xxd.1\fR
.br
0000028: 3235 7468 204d 6179 2031 3939  25th May 199
.PP
.br
Create a 65537 byte file with all bytes 0x00,
except for the last one which is 'A' (hex 0x41).
.br
\fI% echo '010000: 41' | xxd \-r \> file\fR
.PP
.br
Hexdump this file with autoskip.
.br
\fI% xxd \-a \-c 12 file\fR
.br
0000000: 0000 0000 0000 0000 0000 0000  ............
.br
*
.br
000fffc: 0000 0000 40                   ....A
.PP
Create a 1 byte file containing a single 'A' character.
The number after '\-r \-s' adds to the linenumbers found in the file;
in effect, the leading bytes are suppressed.
.br
\fI% echo '010000: 41' | xxd \-r \-s \-0x10000 \> file\fR
.PP
Use xxd as a filter within an editor such as
.B vim(1)
to hexdump a region marked between `a' and `z'.
.br
\fI:'a,'z!xxd\fR
.PP
Use xxd as a filter within an editor such as
.B vim(1)
to recover a binary hexdump marked between `a' and `z'.
.br
\fI:'a,'z!xxd \-r\fR
.PP
Use xxd as a filter within an editor such as
.B vim(1)
to recover one line of a hexdump.  Move the cursor over the line and type:
.br
\fI!!xxd \-r\fR
.PP
Read single characters from a serial line
.br
\fI% xxd \-c1 < /dev/term/b &\fR
.br
\fI% stty < /dev/term/b \-echo \-opost \-isig \-icanon min 1\fR
.br
\fI% echo \-n foo > /dev/term/b\fR
.PP
.SH "RETURN VALUES"
The following error values are returned:
.TP
0
no errors encountered.
.TP
\-1
operation not supported (
.I xxd \-r \-i
still impossible).
.TP
1
error while parsing options.
.TP
2
problems with input file.
.TP
3
problems with output file.
.TP
4,5
desired seek position is unreachable.
.SH "SEE ALSO"
uuencode(1), uudecode(1), patch(1)
.br
.SH WARNINGS
The tools weirdness matches its creators brain.
Use entirely at your own risk. Copy files. Trace it. Become a wizard.
.br
.SH VERSION
This manual page documents xxd version 1.7
.SH AUTHOR
.br
(c) 1990-1997 by Juergen Weigert
.br
<jnweiger@informatik.uni-erlangen.de>
.LP
Distribute freely and credit me,
.br
make money and share with me,
.br
lose money and don't ask me.
.PP
Manual page started by Tony Nugent
.br
<tony@sctnugen.ppp.gu.edu.au> <T.Nugent@sct.gu.edu.au>
.br
Small changes by Bram Moolenaar.
Edited by Juergen Weigert.
.PP


CHANGELOG.rst
Changelog
=========

0.4.0 (TBD)
-----------

* Library rewritten from scratch (not backwards compatible).
* Added new object oriented API, hopefully more user friendly.
* Added *Texas Instruments TI-TXT* file format.
* Improved docs and examples.


0.3.1 (2024-01-23)
------------------

* Added support for Python 3.12.
* Added Motorola header editing.
* Minor fixes and changes.


0.3.0 (2023-02-21)
------------------

* Added support for Python 3.11, removed 3.6.
* Deprecated ``hexrec.blocks`` module entirely.
* Using ``bytesparse`` for virtual memory management.
* Improved repository layout.
* Improved testing and packaging workflow.
* Minor fixes and changes.


0.2.3 (2021-12-30)
------------------

* Removed dependency of legacy pathlib package; using Python's own module instead.
* Added support for Python 3.10.
* Fixed maximum SREC length.
* Changed pattern offset behavior.
* Some alignment to the ``bytesparse.Memory`` API; deprecated code marked as such.


0.2.2 (2020-11-08)
------------------

* Added workaround to register default record types.
* Added support for Python 3.9.
* Fixed insertion bug.
* Added empty space reservation.


0.2.1 (2020-03-05)
------------------

* Fixed flood with empty span.


0.2.0 (2020-02-01)
------------------

* Added support for current Python versions (3.8, PyPy 3).
* Removed support for old Python versions (< 3.6, PyPy 2).
* Major refactoring to allow an easier integration of new record formats.


0.1.0 (2019-08-13)
------------------

* Added support for Python 3.7.


0.0.4 (2018-12-22)
------------------

* New command line interface made with Click.
* More testing and fixing.
* Some refactoring.
* More documentation.


0.0.3 (2018-12-04)
------------------

* Much testing and fixing.
* Some refactoring.
* More documentation.


0.0.2 (2018-08-29)
------------------

* Major refactoring.
* Added most of the documentation.
* Added first drafts to manage blocks of data.
* Added first test suites.


0.0.1 (2018-06-27)
------------------

* First release on PyPI.
* Added first drafts to manage record files.
* Added first emulation of xxd.


src/hexrec/xxd.py
# Copyright (c) 2013-2024, Andrea Zoppi
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

r"""Emulation of the xxd utility."""

import binascii
import io
import os
import re
import sys
from typing import IO
from typing import List
from typing import Optional
from typing import Tuple
from typing import Union

from bytesparse import Memory
from bytesparse.base import ImmutableMemory
from bytesparse.base import MutableMemory

from .base import AnyBytes
from .utils import SparseMemoryIO
from .utils import chop
from .utils import parse_int

_SEEKING_REGEX = re.compile(r'^(?P<sign>\+?-?)-?(?P<absolute>\w+)$')

_REVERSE_REGEX = re.compile(b'^\\s*(?P<address>[A-Fa-f0-9]+)\\s*:\\s*'
                            b'(?P<data>([A-Fa-f0-9]{2}\\s?)+)'
                            b'(\\s.*)?$')

ZERO_BLOCK_SIZE = 1 << 20  # 1 MiB

_HEX_LOWER = [b'%02x' % b for b in range(256)] + [b'--', b'>>', b'<<']
_HEX_UPPER = [b'%02X' % b for b in range(256)] + [b'--', b'>>', b'<<']

_BIN8: List[bytes] = (
    [bin(i)[2:].zfill(8).encode() for i in range(256)] +
    [b'--------', b'>>>>>>>>', b'<<<<<<<<']
)

CHAR_ASCII = (
    b'................'
    b'................'
    b' !"#$%&\'()*+,-./'
    b'0123456789:;<=>?'
    b'@ABCDEFGHIJKLMNO'
    b'PQRSTUVWXYZ[\\]^_'
    b'`abcdefghijklmno'
    b'pqrstuvwxyz{|}~.'
    b'................'
    b'................'
    b'................'
    b'................'
    b'................'
    b'................'
    b'................'
    b'................'
    b' ><'
)
r"""Mapping from integer to ASCII characters."""

CHAR_EBCDIC = (
    b'................'
    b'................'
    b'................'
    b'................'
    b' ...........<(+|'
    b'&.........!$*);~'
    b'-/.........,%_>?'
    b'.........`:#@\'="'
    b'.abcdefghi......'
    b'.jklmnopqr^.....'
    b'..stuvwxyz...[..'
    b'.............]..'
    b'{ABCDEFGHI......'
    b'}JKLMNOPQR......'
    b'\\.STUVWXYZ......'
    b'0123456789......'
    b' ><'
)
r"""Mapping from integer to EBCDIC characters."""


def parse_seek(value: Optional[str]) -> Tuple[str, int]:
    r"""Parses the seek option string.

    Argument:
        value (str):
            The value to convert.
            It is converted to :obj:`str` before processing.
            ``None`` equals zero.

    Returns:
        tuple: ``(sign_string, unsigned_value)``.
    """

    if value is None:
        return '', 0
    else:
        m = _SEEKING_REGEX.match(str(value))
        if not m:
            raise ValueError('invalid seeking')
        ss, sv = m.groups()
        sv = parse_int(sv)
        return ss, sv


def _unhexlify(line: AnyBytes) -> Union[AnyBytes, ImmutableMemory]:

    line = line.translate(None, b' \t.:\r\n')

    if 0x3C in line:
        line = line.replace(b'<', b'-')

    if 0x3E in line:
        line = line.replace(b'>', b'-')

    if 0x2D in line:
        zeroed = line.replace(b'-', b'0')
        chunk = binascii.unhexlify(zeroed)
        even = line[::2]
        size = len(chunk)
        memory = Memory.from_values(((None if even[i] == 0x2D else chunk[i])
                                     for i in range(size)),
                                    start=0, endex=size)
        return memory
    else:
        chunk = binascii.unhexlify(line)
        return chunk


def _write_zeros(
    stream: Union[IO, SparseMemoryIO],
    size: int
) -> None:

    zero_block = bytes(ZERO_BLOCK_SIZE)
    for _ in range(size // ZERO_BLOCK_SIZE):
        stream.write(zero_block)
    del zero_block
    stream.write(bytes(size % ZERO_BLOCK_SIZE))


def xxd_core(
    infile: Optional[Union[str, AnyBytes, IO, ImmutableMemory]] = None,
    outfile: Optional[Union[str, AnyBytes, IO]] = None,
    autoskip: bool = False,
    bits: Optional[int] = None,
    cols: Optional[int] = None,
    ebcdic: bool = False,
    endian: bool = False,
    groupsize: Optional[int] = None,
    include: bool = False,
    length: Optional[int] = None,
    linesep: Optional[bytes] = None,
    offset: Optional[int] = None,
    postscript: bool = False,
    quadword: bool = False,
    revert: bool = False,
    oseek: Optional[int] = None,
    iseek: Optional[Union[int, str]] = None,
    upper_all: bool = False,
    upper: bool = False,
    oseek_zeroes: bool = True,
) -> IO:
    r"""Emulation of the `xxd` utility core.

    Args:
        infile (str or bytes):
            Input data.
            If :obj:`str`, it is considered as the input file path.
            If :obj:`bytes`, it is the input byte chunk.
            If ``None``, it reads from the standard input.

        outfile (str or bytes):
            Output data.
            If :obj:`str`, it is considered as the output file path.
            If :obj:`bytes`, it is the output byte chunk.
            If ``None``, it writes to the standard output.

        autoskip (bool):
            Toggles autoskip. A single ``'*'`` replaces null lines.

        bits (bool):
            Switches to bits (binary digits) dump, rather than hexdump.
            This option writes octets as eight digits of '1' and '0' instead
            of a normal hexadecimal dump.
            Each line is preceded by a line number in hexadecimal and
            followed by an ASCII (or EBCDIC) representation.
            The argument switches ``revert``, ``postscript``, ``include`` do
            not work with this mode.

        cols (int):
            Formats ``cols`` octets per line. Max 256.
            Defaults: normal 16, ``include`` 12, ``postscript`` 30, ``bits`` 6.

        ebcdic (bool):
            Changes the character encoding in the right-hand column from
            ASCII to EBCDIC.
            This does not change the hexadecimal representation.
            The option is meaningless in combinations with ``revert``,
            ``postscript`` or ``include``.

        endian (bool):
            Switches to little-endian hexdump.
            This option treats  byte groups as words in little-endian byte
            order.
            The default grouping of 4 bytes may be changed using ``groupsize``.
            This option only applies to hexdump, leaving the ASCII (or EBCDIC)
            representation unchanged.
            The switches ``revert``, ``postscript``, ``include`` do not work
            with this mode.

        groupsize (int):
            Separates the output of every ``groupsize``
            bytes (two hex characters or eight bit-digits each) by a whitespace.
            Specify ``groupsize`` 0 to suppress grouping.
            ``groupsize`` defaults to 2 in normal mode, 4 in little-endian mode
            and 1 in bits mode. Grouping does not apply to ``postscript`` or
            ``include``.

        include (bool):
            Output in C include file style.
            A complete static array definition is written (named after the
            input file), unless reading from standard input.

        length (int):
            Stops after writing ``length`` octets.

        linesep (bytes):
            Line separator characters.
            If ``None``, it defaults to ``os.linesep.encode()``.

        offset (int):
            Adds ``offset`` to the displayed file position.

        postscript (bool):
            Outputs in postscript continuous hexdump style.
            Also known as plain hexdump style.

        quadword (bool):
            Uses 64-bit addressing.

        revert (bool):
            Reverse operation: convert (or patch) hexdump into binary.
            If not writing to standard output, it writes into its output file
            without truncating it.
            Use the combination ``revert`` and ``postscript`` to read plain
            hexadecimal dumps without line number information and without a
            particular column layout.
            Additional Whitespace and line breaks are allowed anywhere.

        oseek (int):
            When used after ``revert`` reverts with ``offset`` added to file
            positions found in hexdump.

        iseek (int or str):
            Starts at ``iseej`` bytes absolute (or relative) input offset.
            Without ``iseek`` option, it starts at the current file position.
            The prefix is used to compute the offset.
            ``+`` indicates that the seek is relative to the current input
            position.
            ``-`` indicates that the seek should be that many characters from
            the end of the input.
            ``+-`` indicates that the seek should be that many characters
            before the current stdin file position.

        upper_all (bool):
            Uses upper case hex letters on address and data.

        upper (bool):
            Uses upper case hex letters on data only.

        oseek_zeroes (bool):
            Output seeking fills with zeros.
            Only affects `outfile` of :class:`bytesparse.base.MutableMemory`.

    Returns:
        stream: The handle to the output stream.
    """
    if cols is not None and not 1 <= cols <= 256:
        raise ValueError('invalid column count')

    if upper_all:
        upper = upper_all

    if (bits or endian) and (postscript or include or revert):
        raise ValueError('incompatible options')

    if sum(bool(_) for _ in [postscript, include, bits]) > 1:
        raise ValueError('incompatible options')

    if not revert and oseek is not None:
        raise ValueError('incompatible options')
    elif oseek is not None and oseek < 0:
        raise ValueError('invalid seeking')

    if linesep is None:
        linesep = os.linesep.encode()

    instream: Optional[IO, SparseMemoryIO] = None
    outstream: Optional[IO, SparseMemoryIO] = None
    outsparse = False

    try:
        # Input stream binding
        if infile is None:
            infile = None
            instream = sys.stdin.buffer
        elif isinstance(infile, str):
            instream = open(infile, 'rb')
        elif isinstance(infile, (bytes, bytearray, memoryview)):
            instream = io.BytesIO(infile)
        elif isinstance(infile, ImmutableMemory):
            instream = SparseMemoryIO(memory=infile)
        else:
            instream = infile

        # Output stream binding
        if outfile is None:
            outfile = None
            outstream = sys.stdout.buffer
        elif isinstance(outfile, str):
            mode = 'r+b' if revert and os.path.exists(outfile) else 'wb'
            outstream = open(outfile, mode)
        elif isinstance(outfile, MutableMemory):
            outstream = SparseMemoryIO(memory=outfile)
            outsparse = True
        else:
            outstream = outfile

        # Input seeking
        offset = parse_int(offset or 0)

        if iseek is not None:
            ss, sv = parse_seek(str(iseek))

            if revert:
                if '-' in ss:
                    sv = -sv
                iseek = sv

            else:
                if ss == '+':
                    instream.seek(sv, io.SEEK_CUR)
                elif ss == '+-':
                    instream.seek(-sv, io.SEEK_CUR)
                elif ss == '-':
                    instream.seek(-sv, io.SEEK_END)
                else:  # elif ss == '':
                    instream.seek(sv, io.SEEK_SET)

            offset += instream.tell()

        # Output seeking
        if revert:
            if outsparse and not oseek_zeroes:
                outstream.seek((oseek or 0), io.SEEK_CUR)
            else:
                _write_zeros(outstream, (oseek or 0))

        # Output mode handling
        if revert:
            if postscript:
                # Plain hexadecimal input
                for line in instream:
                    data = _unhexlify(line)
                    outstream.write(data)
            else:
                if cols is None:
                    cols = 16

                for line in instream:
                    match = _REVERSE_REGEX.match(line)
                    if match:
                        # Interpret line contents
                        groups = match.groupdict()
                        address = (oseek or 0) + (iseek or 0) + int(groups['address'], 16)
                        data = _unhexlify(groups['data'])
                        data = data[:cols]

                        # Write line data (fill gaps if needed)
                        outstream.seek(0, io.SEEK_END)
                        outoffset = outstream.tell()
                        if outoffset < address:
                            if outsparse and not oseek_zeroes:
                                outstream.seek((address - outoffset), io.SEEK_CUR)
                            else:
                                _write_zeros(outstream, (address - outoffset))
                        outstream.seek(address, io.SEEK_SET)
                        outstream.write(data)

            raise StopIteration  # End of input stream

        elif postscript:
            # Plain hexadecimal output
            if cols is None:
                cols = 30

            count = 0
            while True:
                if length is None:
                    chunk = instream.read(cols)
                else:
                    chunk = instream.read(min(cols, length - count))

                if chunk:
                    table = _HEX_UPPER if upper else _HEX_LOWER
                    line = b''.join(table[b] for b in chunk)
                    outstream.write(line)
                    outstream.write(linesep)
                    count += len(chunk)
                else:
                    raise StopIteration  # End of input stream

        elif bits:
            if cols is None:
                cols = 6
            if groupsize is None:
                groupsize = 1

        elif include:
            if cols is None:
                cols = 12

            # Data variable definition
            if isinstance(infile, str):
                label = os.path.basename(infile)
                label = re.sub('[^0-9a-zA-Z]+', '_', label).encode()
                outstream.write(b'unsigned char %s[] = {%s' % (label, linesep))
            else:
                label = None

            indent = b'  0X' if upper_all else b'  0x'
            sep = b', 0X' if upper_all else b', 0x'

            count = 0
            comma_linesep = b',' + linesep
            while True:
                if length is None:
                    chunk = instream.read(cols)
                else:
                    chunk = instream.read(min(cols, length - count))

                if chunk:
                    if count:
                        outstream.write(comma_linesep)
                    outstream.write(indent)
                    table = _HEX_UPPER if upper else _HEX_LOWER
                    text = sep.join(table[b] for b in chunk)
                    outstream.write(text)
                    count += len(chunk)

                else:
                    # Data end and length variable definition
                    if isinstance(infile, str):
                        outstream.write(b'%s};%sunsigned int %s_len = %d;%s'
                                        % (linesep, linesep, label, count, linesep))
                    else:
                        outstream.write(linesep)

                    raise StopIteration  # End of input stream

        else:
            if cols is None:
                cols = 16

        if groupsize is None:
            groupsize = 4 if endian else 2
        if not 0 <= groupsize <= 256:
            raise ValueError('invalid grouping')

        data_width = (cols * (8 if bits else 2) +
                      ((cols - 1) // groupsize if groupsize else 0))

        line_fmt = b'%%0%d%s: %%-%ds  %%s%s' % (
            (16 if quadword else 8),
            (b'X' if upper_all else b'x'),
            data_width,
            linesep
        )

        # Hex dump
        if not 0 <= offset < 0xFFFFFFFF:
            raise ValueError('offset overflow')

        last_zero = None
        count = 0

        while True:
            # Input byte columns
            if length is None:
                chunk = instream.read(cols)
            else:
                chunk = instream.read(min(cols, length - count))

            if chunk:
                # Null line skipping
                if autoskip and all(b == 0 for b in chunk):
                    if last_zero:
                        offset += len(chunk)
                        count += len(chunk)
                        continue
                    else:
                        last_zero = Ellipsis

                # Byte grouping
                if groupsize:
                    tokens = chop(chunk, groupsize)
                else:
                    tokens = [chunk]

                if bits:
                    table = _BIN8
                    tokens = b' '.join(b''.join(table[b] for b in t) for t in tokens)
                elif groupsize:
                    table = _HEX_UPPER if upper else _HEX_LOWER
                    if endian:
                        tokens = b' '.join(b''.join(table[b] for b in reversed(t)) for t in tokens)
                    else:
                        tokens = b' '.join(b''.join(table[b] for b in t) for t in tokens)
                else:
                    table = _HEX_UPPER if upper else _HEX_LOWER
                    tokens = b' '.join(b''.join(table[b] for b in t) for t in tokens)

                # Comment text generation
                charset = CHAR_EBCDIC if ebcdic else CHAR_ASCII
                text = bytes(charset[b] for b in chunk)

                # Line output
                line = line_fmt % (offset, tokens, text)
                outstream.write(line)

                offset += len(chunk)
                count += len(chunk)

                if last_zero is Ellipsis:
                    last_zero = True
                    outstream.write(b'*')
                    outstream.write(linesep)

            else:
                raise StopIteration  # End of input stream

    except StopIteration:
        pass

    finally:
        if instream is not None and isinstance(infile, str):
            instream.close()

        if outstream is not None and isinstance(outfile, str):
            outstream.close()

    return outstream
